uber/site_sections/registration.py

```diff
@@ -217,7 +217,10 @@ def change_badge(self, session, id, message='', **params):
         }
 
     def promo_code_groups(self, session, message=''):
-        groups = session.query(PromoCodeGroup).options(joinedload('buyer')).order_by(PromoCodeGroup.name).all()
+        groups = session.query(PromoCodeGroup).options(
+                                                joinedload(PromoCodeGroup.buyer),
+                                                joinedload(PromoCodeGroup.promo_codes)
+                                                ).order_by(PromoCodeGroup.name).all()
```

Comment:
From looking through the template, we're doing

```
{{ group.promo_codes|length }}
```

as our way of counting the codes for each group.  However, this approach requires us to load all of the promo codes just to count them.  This query itself is basically the entire reason for this page load time being high:

```python
>>> before = time(); groups = session.query(PromoCodeGroup).options(joinedload(PromoCodeGroup.buyer), joinedload(PromoCodeGroup.promo_codes)).order_by(PromoCodeGroup.name).all(); after = time(); print(after - before)
3.824937582015991
```

Since you've already implemented a ``used_counts`` option at https://github.com/magfest/ubersystem/pull/3675/files then we can probably just eliminate the ``joinedload(PromoCodeGroup.promo_codes)`` since it will no longer be needed, and then the page will speed up by 1-2 orders of magnitude.

Reply:
Oh, actually there would need to be one additional dict as well, because of

```
{{ group.total_cost }}
```

which iterates through the promo codes, so we can also do

```python
total_costs = {
    group_id: total for group_id, total in
        session.query(PromoCode.group_id, func.sum(PromoCode.cost))
               .group_by(PromoCode.group_id)
}
```

and then we can do a lookup without needing the ``joinedload``.